### PR TITLE
Enable custom unitary operations in Python for simulation

### PR DESCRIPTION
**Pull request by khalatepradnya**
Python support for custom operations using new API - `cudaq.register_operation`

This first PR focuses solely on Python frontend and simulation, and follow up PRs will cover parameterized operations and decomposition (synthesis)

### CODE REVIEW

**Review by schweitzpgi — CHANGES_REQUESTED** [review]
I don't see a C++ test or changes to the C++ bridge.

Is unitary circuit synthesis part of another PR?

**Comment by khalatepradnya** [comment]
Thanks, Eric, for the comments. I will address these.
The C++ MLIR support will be part of a separate PR but will depend on this, like `QuakeOps`.

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td [hunk 1]
```diff
@@ -1157,4 +1157,38 @@ def ZOp : OneTargetOp<"z", [Hermitian]> {
...
+    compile-time known attribute, or a runtime value. 
+
+    If the unitary is provided at compile time, all target operands must 
+    have constant size (e.g. veq<?> will result in an error).
+
+    For compile-time unitary data, we represent as an ArrayAttr of DenseF32ArrayAttr 
+    elements, each element is a 2-tuple, the real and imaginary parts.
+  }];
+
+  let arguments = (ins StrAttr:$opName,
```
[the remark this answers is not in the document]
Comment (khalatepradnya) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:1176]: Will be addressed in https://github.com/NVIDIA/cuda-quantum/pull/1794.
I will update this after that PR.

**Comment by khalatepradnya** [comment]
Covered by https://github.com/NVIDIA/cuda-quantum/pull/1794

**Pull request closed without merging**